src/client/GraphQLClient/client.ts

```
@@ -609,6 +650,110 @@ export class GitHubGraphQLClient {
+          endpoint: `retryGraphQL`,
+        });
+      }
+      validateGraphQLResponse(response, logger, queryString);
+      return response;
+    };
+
+    // Check https://github.com/lifeomic/attempt for options on retry
+    return await retry(queryWithRateLimitCatch, {
+      maxAttempts: 3,
+      delay: 30_000, // 30 seconds to start
+      timeout: 180_000, // 3 min timeout. We need this in case Node hangs with ETIMEDOUT
+      factor: 2, //exponential backoff factor. with 30 sec start and 3 attempts, longest wait is 2 min
+      handleError: async (err, attemptContext) => {
+        /* retry will keep trying to the limits of retryOptions
+         * but it lets you intervene in this function - if you throw an error from in here,
+         * it stops retrying. Otherwise you can just log the attempts.
+         *
+         * Github has "Secondary Rate Limits" in case of excessive polling or very costly API calls.
+         * GitHub guidance is to "wait a few minutes" when we get one of these errors.
+         * https://docs.github.com/en/rest/overview/resources-in-the-rest-api#secondary-rate-limits
+         * this link is REST specific - however, the limits might apply to GraphQL as well,
+         * and our GraphQL client is not using the @octokit throttling and retry plugins like our REST client
+         * therefore some retry logic is appropriate here
+         */
+
+        if (err.status === 401) {
+          logger.warn(
+            { attemptContext, err },
+            `Hit 401, attempting to refresh the token and try again.`,
+          );
+          await this.refreshToken();
+          return;
+        }
+
+        // don't keep trying if it's not going to get better
+        if (err.retryable === false || err.status === 403) {
+          logger.warn(
+            { attemptContext, err },
+            `Hit an unrecoverable error when attempting to query GraphQL. Aborting.`,
+          );
+          attemptContext.abort();
```

So a 403 is the only error we abort and give up?  Would we also want to add a 500 error catch at a minimum just to catch the edge case where the server is falling on its face?  I don't imagine it's a huge issue if we retry in that instance, but it'd let us fail faster.

---

I copy/pasted this method a changed a very small portion of it. It seems to be working as I would hope so I'll go back and clean this up. I'll also consider the scenario you presented here.